src/noit_filters_lmdb.c

```diff
@@ -1569,39 +1664,33 @@ noit_filters_lmdb_rest_set_filter(mtev_http_rest_closure_t *restc,
     error_code = 406;
     goto error;
   }
-  
-  exists = noit_filters_lmdb_already_in_db(pats[1]);
-  if (exists) {
-    old_seq = noit_filters_lmdb_get_seq(pats[1]);
-  }
 
-  if((newfilter = noit_filter_validate_filter(indoc, pats[1], &seq, &error)) == NULL) {
-    goto error;
-  }
+  mtev_http_connection *connection = mtev_http_session_connection(ctx);
+  mtevAssert(connection);
 
-  if(exists && (old_seq >= seq && seq != 0)) {
-    error = "sequencing error";
-    error_code = 409;
-    goto error;
-  }
-  error_str = NULL;
-  rv = noit_filters_lmdb_convert_one_xml_filterset_to_lmdb(mtev_conf_section_from_xmlnodeptr(newfilter), &error_str);
-  if (rv || error_str) {
-    mtevL(mtev_error, "noit_filters_lmdb_process_repl: error converting filterset from xml to lmdb: %s\n", error_str ? error_str : "(unknown error)");
-    error_code = 500;
-    error = error_str;
-    allocated_error = mtev_true;
-    goto error;
-  }
-  noit_filter_compile_add(mtev_conf_section_from_xmlnodeptr(newfilter));
+  eventer_t conne = mtev_http_connection_event_float(connection);
+  if(conne) eventer_remove_fde(conne);
 
-  if(restc->call_closure_free) {
-    restc->call_closure_free(restc->call_closure);
-  }
-  restc->call_closure_free = NULL;
-  restc->call_closure = NULL;
-  restc->fastpath = noit_filters_lmdb_rest_show_filter;
-  return restc->fastpath(restc, restc->nparams, restc->params);
+  lmdb_filter_set_closure_t *lfsc = (lmdb_filter_set_closure_t *)calloc(1, sizeof(*lfsc));
+  lfsc->restc = restc;
+  lfsc->indoc = indoc;
+  lfsc->name = strdup(pats[1]);
+  /* rest_get_xml_upload sets a closure and free function... we need to set
+   * our own closure, though, so we need to save off what was there and call
+   * it when we clean up */
+  lfsc->xml_data = restc->call_closure;
+  lfsc->xml_data_free = restc->call_closure_free;
+
+  restc->call_closure = lfsc;
+  restc->call_closure_free = lmdb_set_filter_data_free;
+  restc->fastpath = noit_filters_lmdb_rest_set_filter_complete;
+
+  eventer_t e = eventer_alloc_asynch(noit_filters_lmdb_rest_set_filter_asynch, lfsc);
+  if(conne) eventer_set_owner(e, eventer_get_owner(conne));
```

Comment:
I don't believe this is sufficient to only set the owner by the HTTP connection.   

```c++
noit_filters_lmdb_already_in_db(lfsc->name);
```

Calls like this one could still produce incorrect results if they're being called by multiple threads at the same time using the same value for the name, I believe the owner must be selected by the name in this case to avoid this

Reply:
@pjulien - Not sure what you mean by "owner must be selected by the name". What do you suggest?

Reply:
``eventer_set_owner`` selects a thread to run in the new pool.  I do not know much about this code but it seems it would be possible coming in from multiple threads that have the same value for the value ``name``

This could break the state management here

Reply:
Temporarily fixed by setting the jobq concurrency down to 1 and disallowing multiple threads running filterset sets.